*** File.pm.dist        Mon Jul 21 11:31:50 2003
--- File.pm     Mon Jul 21 11:33:00 2003
***************
*** 133,139 ****
      my @files = readdir(DIR);
      foreach my $file (@files) {
          if ($file =~ /^Apache-Session.*\.lock$/) {
!             if ((stat($dir.'/'.$file))[8] - $now >= $time) {
                  open(FH, "+>$dir/".$file) || next;
                  flock(FH, LOCK_EX) || next;
                  unlink($dir.'/'.$file) || next;
--- 133,139 ----
      my @files = readdir(DIR);
      foreach my $file (@files) {
          if ($file =~ /^Apache-Session.*\.lock$/) {
!             if ($now - (stat($dir.'/'.$file))[8] >= $time) {
                  open(FH, "+>$dir/".$file) || next;
                  flock(FH, LOCK_EX) || next;
                  unlink($dir.'/'.$file) || next;

